src/Router.js

```diff
@@ -85,62 +70,119 @@ export default class Router {
       throw new Error('Router#registerRoutes must be passed an array of Routes')
     }
     routes.map(route => {
-      this.__routes.push(new Route(route))
+      route = new Route(route);
+      if (this.onRouteComplete) {
+        const handlerLength = route.handlers.length-1;
+        const lastHandler = route.handlers[handlerLength];
+        const routingEnd = (ctx, next) => {
+          lastHandler(ctx);
+          const duration = fns.getNow() - this.__startTime;
+          const fromPath = this.__fromPath || 'PAGE LOAD';
+          this.onRouteComplete({
+            fromPath,
+            toPath: this.__currentCanonicalPath,
+            duration,
+          })
+        };
+        route.handlers = route.handlers.slice(0, handlerLength).concat(routingEnd);
+      }
+      if (this.onRouteStart) {
+        route.handlers.unshift(this.onRouteStart);
+      }
+      this.__routes.push(route);
     })
   }
 
   /**
+   * @param {String} path
+   */
+  registerCatchallPath(path) {
+    this.__catchallPath = path;
+  }
+
+  /**
    * @param {String} canonicalPath
    */
   go(canonicalPath) {
-    this.__dispatch(canonicalPath, true)
+    this.__dispatch(canonicalPath, false)
   }
 
   /**
    * @param {String} canonicalPath
    */
   replace(canonicalPath) {
-    this.__dispatch(canonicalPath, false)
+    this.__dispatch(canonicalPath, true)
   }
 
   reset() {
-    this.__routes = []
+    this.__catchallPath = null;
+    this.__routes = [];
+    WindowEnv.removeEventListener('popstate', this.__onpopstate)
+  }
+
+  catchall() {
+    WindowEnv.navigate(this.__catchallPath)
   }
 
+
   /**
-   * @param {String} canonicalPath
+   * @param {RouterHandler[]} handlers
    * @param {Context} ctx
    */
-  catchall(ctx) {
-    WindowEnv.navigate(ctx.canonicalPath)
+  __runHandlers(handlers, ctx, callback) {
+    let len = handlers.length
+    let i = 0;
+
+    let next = () => {
+      if (this.__currentCanonicalPath !== ctx.canonicalPath) {
```

Comment:
Hmmm i'm a little concerned over this mechanism to escape the route.

what happens if you click a new route, then click back to the same route.  I suppose it's synchronous so it will probably work.

One idea I had was to just keep track of a number / id that way any new dispatch will abondon the old dispatch instead of trying to use the canonical path as the identifier.

Ex

```
constructor() {
  this.dispatchId = 1;
}

__dispatch() {
  this.dispatchId++;
  if (ctx.dispatchId !== this.dispatchId) { 
    return
  }
}
```